File: server/lib/env.ts
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import dotenv from 'dotenv';

let loadedEnvFile: string | null = null;

/** Repo root (folder containing root package.json with npm workspaces). */
export function repoRoot(): string {
  let dir = path.dirname(fileURLToPath(import.meta.url));
  for (let i = 0; i < 8; i++) {
    const pkgPath = path.join(dir, 'package.json');
    if (fs.existsSync(pkgPath)) {
      try {
        const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8')) as { workspaces?: unknown };
        if (pkg.workspaces) return dir;
      } catch {
        // keep walking
      }
    }
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');
}

function walkEnvPaths(start: string): string[] {
  const paths: string[] = [];
  let dir = path.resolve(start);
  for (let i = 0; i < 8; i++) {
    paths.push(path.join(dir, '.env'));
    paths.push(path.join(dir, '.env.txt'));
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return paths;
}

export function envSearchPaths(): string[] {
  const unique = new Set<string>();
  const ordered: string[] = [];

  const add = (p: string) => {
    const resolved = path.resolve(p);
    if (!unique.has(resolved)) {
      unique.add(resolved);
      ordered.push(resolved);
    }
  };

  if (process.env.DOTENV_PATH) add(process.env.DOTENV_PATH);
  for (const p of walkEnvPaths(repoRoot())) add(p);
  for (const p of walkEnvPaths(process.cwd())) add(p);

  return ordered;
}

export function loadEnv(): string | null {
  for (const envPath of envSearchPaths()) {
    if (!fs.existsSync(envPath)) continue;
    const result = dotenv.config({ path: envPath });
    if (!result.error) {
      loadedEnvFile = envPath;
      console.log(`[env] loaded ${envPath}`);
      return envPath;
    }
    console.warn(`[env] failed to parse ${envPath}: ${result.error.message}`);
  }

  loadedEnvFile = null;
  console.warn(`[env] no .env found. repo_root=${repoRoot()} cwd=${process.cwd()}`);
  return null;
}

export function getLoadedEnvFile(): string | null {
  return loadedEnvFile;
}

export function sqlEnvStatus(): {
  repo_root: string;
  cwd: string;
  env_file: string | null;
  env_search: Array<{ path: string; exists: boolean }>;
  missing: string[];
  host_set: boolean;
  token_set: boolean;
  warehouse_set: boolean;
} {
  const host = process.env.DATABRICKS_HOST || process.env.DATABRICKS_SERVER_HOSTNAME || '';
  const token = process.env.DATABRICKS_PAT_TOKEN || process.env.DATABRICKS_TOKEN || '';
  const warehouse = process.env.DATABRICKS_WAREHOUSE_ID || '';

  const missing: string[] = [];
  if (!host.trim()) missing.push('DATABRICKS_HOST or DATABRICKS_SERVER_HOSTNAME');
  if (!token.trim()) missing.push('DATABRICKS_PAT_TOKEN');
  if (!warehouse.trim()) missing.push('DATABRICKS_WAREHOUSE_ID');

  const search = envSearchPaths().map(p => ({ path: p, exists: fs.existsSync(p) }));

  return {
    repo_root: repoRoot(),
    cwd: process.cwd(),
    env_file: getLoadedEnvFile() ?? search.find(s => s.exists)?.path ?? null,
    env_search: search,
    missing,
    host_set: Boolean(host.trim()),
    token_set: Boolean(token.trim()),
    warehouse_set: Boolean(warehouse.trim()),
  };
}

File: py_server/lib/env.py
"""Environment loading — mirrors server/lib/env.ts."""
from __future__ import annotations

import json
import os
from pathlib import Path

from dotenv import load_dotenv

_loaded_env_file: str | None = None


def repo_root() -> str:
    """Repo root (folder containing root package.json with npm workspaces)."""
    dir_path = Path(__file__).resolve().parent
    for _ in range(8):
        pkg_path = dir_path / "package.json"
        if pkg_path.is_file():
            try:
                pkg = json.loads(pkg_path.read_text(encoding="utf-8"))
                if pkg.get("workspaces"):
                    return str(dir_path)
            except (json.JSONDecodeError, OSError):
                pass
        parent = dir_path.parent
        if parent == dir_path:
            break
        dir_path = parent
    return str(Path(__file__).resolve().parent.parent.parent)


def _walk_env_paths(start: str) -> list[str]:
    paths: list[str] = []
    dir_path = Path(start).resolve()
    for _ in range(8):
        paths.append(str(dir_path / ".env"))
        paths.append(str(dir_path / ".env.txt"))
        parent = dir_path.parent
        if parent == dir_path:
            break
        dir_path = parent
    return paths


def env_search_paths() -> list[str]:
    unique: set[str] = set()
    ordered: list[str] = []

    def add(p: str) -> None:
        resolved = str(Path(p).resolve())
        if resolved not in unique:
            unique.add(resolved)
            ordered.append(resolved)

    dotenv_path = os.environ.get("DOTENV_PATH")
    if dotenv_path:
        add(dotenv_path)
    for p in _walk_env_paths(repo_root()):
        add(p)
    for p in _walk_env_paths(os.getcwd()):
        add(p)

    return ordered


def load_env() -> str | None:
    global _loaded_env_file
    for env_path in env_search_paths():
        path = Path(env_path)
        if not path.is_file():
            continue
        try:
            load_dotenv(env_path, override=False)
            _loaded_env_file = env_path
            print(f"[env] loaded {env_path}")
            return env_path
        except Exception as exc:
            print(f"[env] failed to parse {env_path}: {exc}")

    _loaded_env_file = None
    print(f"[env] no .env found. repo_root={repo_root()} cwd={os.getcwd()}")
    return None


def get_loaded_env_file() -> str | None:
    return _loaded_env_file


def sql_env_status() -> dict:
    host = os.environ.get("DATABRICKS_HOST") or os.environ.get("DATABRICKS_SERVER_HOSTNAME") or ""
    token = os.environ.get("DATABRICKS_PAT_TOKEN") or os.environ.get("DATABRICKS_TOKEN") or ""
    warehouse = os.environ.get("DATABRICKS_WAREHOUSE_ID") or ""

    missing: list[str] = []
    if not host.strip():
        missing.append("DATABRICKS_HOST or DATABRICKS_SERVER_HOSTNAME")
    if not token.strip():
        missing.append("DATABRICKS_PAT_TOKEN")
    if not warehouse.strip():
        missing.append("DATABRICKS_WAREHOUSE_ID")

    search = [{"path": p, "exists": Path(p).is_file()} for p in env_search_paths()]

    return {
        "repo_root": repo_root(),
        "cwd": os.getcwd(),
        "env_file": get_loaded_env_file() or next((s["path"] for s in search if s["exists"]), None),
        "env_search": search,
        "missing": missing,
        "host_set": bool(host.strip()),
        "token_set": bool(token.strip()),
        "warehouse_set": bool(warehouse.strip()),
    }
